builder/vmware/iso/step_create_vmx.go

```diff
@@ -155,7 +156,34 @@ func (s *stepCreateVMX) Run(state multistep.StateBag) multistep.StepAction {
 	return multistep.ActionContinue
 }
 
-func (s *stepCreateVMX) Cleanup(multistep.StateBag) {
+func (s *stepCreateVMX) Cleanup(state multistep.StateBag) {
+	driver := state.Get("driver").(vmwcommon.Driver)
+	config := state.Get("config").(*Config)
+	ui := state.Get("ui").(packer.Ui)
+
+	ui.Say("Unregistering and deleting virtual machine...")
+	var err error = nil
+	for i := 0; i < 5; i++ {
+		running, err := driver.IsRunning(config.VMXTemplatePath)
+		if err == nil {
+			if running {
+				err = driver.Stop(config.VMXTemplatePath)
+			}
+		}
+		if err == nil {
```

Comment:
Sorry! I'm getting confused about what `err` is and when it should be set. The intent of the code is not clear to me, so I think it's a [code smell](http://c2.com/cgi/wiki?CodeSmell).

Go generally uses `return err` to indicate a failure and since we can't do that inside the loop I think it may be clearer to put that logic in a function. Something like this?

```
func DestroyVM(driver vmwcommon.Driver, vmxpath string) error {
	running, err := driver.IsRunning(vmxpath)
	if err != nil {
		return err
	}

	if running {
		err = driver.Stop(vmxpath)
		if err != nil {
			return err
		}
		// Wait a few seconds for the VM to stop
		time.Sleep(5 * time.Second)
	}

	err = driver.Delete(vmxpath)
	if err == nil {
		return err
	}

	return nil
}

...

	var err error = nil
	for i := 0; i < 5; i++ {
		err = DestroyVM(config.VMXTemplatePath)
		if err == nil {
			break
		}
	}

	if err != nil {
		ui.Error(fmt.Sprintf("Error deleting virtual machine: %s", err))
	}
```

I can see that each check against the VM state returns an error and exits the function if there is a problem (but I don't confuse this with exiting the loop). I can also write a unit test around this function more easily.

With this approach you lose the increasing retry time but I think the code is clearer overall. What are your thoughts?